File: java/org.financer.client.javafx/src/test/java/org/financer/client/javafx/main/AbstractFinancerApplicationTest.java
package org.financer.client.javafx.main;

import com.jfoenix.controls.*;
import javafx.scene.Node;
import javafx.scene.control.Button;
import javafx.scene.control.Label;
import javafx.scene.control.TextField;
import javafx.scene.input.KeyCode;
import javafx.scene.input.MouseButton;
import org.financer.client.domain.model.category.Category;
import org.financer.client.domain.model.transaction.*;
import org.financer.client.domain.model.user.Token;
import org.financer.client.domain.model.user.User;
import org.financer.client.domain.model.user.VerificationToken;
import org.financer.client.format.I18N;
import org.financer.client.javafx.components.DatePicker;
import org.financer.client.javafx.components.DoubleField;
import org.financer.client.javafx.components.IntegerField;
import org.financer.client.javafx.format.JavaFXFormatter;
import org.financer.client.javafx.local.LocalStorageImpl;
import org.financer.shared.domain.model.value.objects.*;
import org.financer.util.collections.TreeUtil;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.testfx.api.FxToolkit;
import org.testfx.framework.junit5.ApplicationTest;

import java.time.LocalDate;
import java.util.Collections;
import java.util.Comparator;
import java.util.HashSet;
import java.util.concurrent.TimeoutException;
import java.util.function.Predicate;

public abstract class AbstractFinancerApplicationTest extends ApplicationTest {

    static final int SHORT_SLEEP = 500;
    static final int MEDIUM_SLEEP = 1000;
    private static final int LONG_SLEEP = 5000;

    JavaFXFormatter formatter;

    static {
        System.setProperty("testfx.robot", "glass");
        System.setProperty("glass.platform", "Monocle");
        System.setProperty("monocle.platform", "Headless");
    }

    protected User user() {
        return new User()
                .setId(1)
                .setEmail(new Email("test@test.com"))
                .setName(new Name("Test", "User"))
                .setPassword(new HashedPassword(password()))
                .setTokens(new HashSet<>(Collections.singletonList(token())))
                .setVerified(false);
    }

    protected String password() {
        return "password";
    }

    protected Token token() {
        return new Token()
                .setId(1)
                .setExpireDate(new ExpireDate())
                .setIpAddress(new IPAddress("192.168.0.1"))
                .setToken(tokenString())
                .setOperatingSystem(new OperatingSystem(OperatingSystem.Values.LINUX));
    }

    protected VerificationToken verificationToken() {
        return new VerificationToken()
                .setId(1)
                .setExpireDate(new ExpireDate(LocalDate.now().plusDays(15)))
                .setToken(tokenString());
    }

    protected TokenString tokenString() {
        return new TokenString("Z6XCS3tyyBlhPfsv7rMxLgfdyEOlUkv0CWSdbFYHCNX2wLlwpH97lJNP69Bny3XZ");
    }

    protected Category variableCategory() {
        return new Category()
                .setId(1)
                .setUser(user())
                .setCategoryClass(new CategoryClass(CategoryClass.Values.VARIABLE_EXPENSES))
                .setName("Variable Category")
                .setParent(null);
    }

    protected Category variableCategoryParent() {
        return new Category()
                .setId(2)
                .setUser(user())
                .setCategoryClass(new CategoryClass(CategoryClass.Values.VARIABLE_EXPENSES))
                .setName("Variable Category Parent")
                .setParent(null);
    }

    protected Category fixedCategory() {
        return new Category()
                .setId(2)
                .setUser(user())
                .setCategoryClass(new CategoryClass(CategoryClass.Values.FIXED_EXPENSES))
                .setName("Fixed Category")
                .setParent(null);
    }

    protected Product product() {
        return new Product()
                .setId(1)
                .setName("Test Product")
                .setQuantity(new Quantity(2))
                .setAmount(new Amount(50));
    }

    protected VariableTransaction variableTransaction() {
        return new VariableTransaction()
                .setId(1)
                .setValueDate(new ValueDate())
                .setCategory(variableCategory())
                .setDescription("Test Purpose")
                .setVendor("Test Vendor")
                .addProduct(product());
    }

    protected Attachment attachment() {
        return new Attachment()
                .setId(1)
                .setName("test.pdf")
                .setTransaction(variableTransaction())
                .setUploadDate(LocalDate.now());
    }

    protected FixedTransactionAmount fixedTransactionAmount() {
        return new FixedTransactionAmount()
                .setId(1)
                .setAmount(new Amount(50))
                .setValueDate(new ValueDate());
    }

    protected FixedTransaction fixedTransaction() {
        return new FixedTransaction()
                .setId(2)
                .setCategory(fixedCategory())
                .setTimeRange(new TimeRange())
                .setIsVariable(false)
                .setAmount(new Amount(50.0))
                .setDay(1)
                .setDescription("Test Purpose")
                .setVendor("Test Vendor")
                .addFixedTransactionAmount(fixedTransactionAmount());
    }

    @BeforeEach
    void setUpEach() throws Exception {
    }

    final <T extends Node> T find(final String query) {
        return lookup(query).query();
    }

    final <T extends Node> T find(Predicate<T> predicate) {
        return lookup(predicate).query();
    }

    void register(User user, String password) {
        clickOn((JFXButton) find("#openRegisterDialogBtn"));

        clickOn((JFXTextField) find("#registerNameTextField"));
        write(user.getName().getFirstName());
        clickOn((JFXTextField) find("#registerSurnameTextField"));
        write(user.getName().getSurname());
        clickOn((JFXTextField) find("#registerEmailTextField"));
        write(user.getEmail().getEmailAddress());
        DatePicker birthDatePicker = find("#registerBirthDatePicker");
        birthDatePicker.setValue(user.getBirthDate().getBirthDate());
        clickOn((JFXComboBox) find("#genderComboBox"));
        press(KeyCode.DOWN).release(KeyCode.DOWN);
        press(KeyCode.ENTER).release(KeyCode.ENTER);
        clickOn((JFXPasswordField) find("#registerPasswordTextField"));
        write(password);
        clickOn((JFXPasswordField) find("#registerRepeatPasswordTextField"));
        write(password);
        confirmDialog();

        sleep(LONG_SLEEP);
        formatter = new JavaFXFormatter(LocalStorageImpl.getInstance());
    }

    void login(User user, String password) {
        clickOn((TextField) find("#loginEmailTextField"));
        write(user.getEmail().getEmailAddress());
        clickOn((TextField) find("#loginPasswordField"));
        write(password);
        clickOn((Button) find("#loginBtn"));
    }

    void logout() throws Exception {
        clickOn(formatter.format(user().getName()));
        clickOn("Logout");
    }

    void addCategory(Category category) {
        clickOn((JFXButton) find("#profileTabBtn"));
        sleep(MEDIUM_SLEEP);

        press(KeyCode.RIGHT).release(KeyCode.RIGHT);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);

        Button newCategoryBtn = find("#newCategoryBtn");
        clickOn(I18N.get(category.getCategoryClass().getCategoryClass().getName()));
        clickOn(newCategoryBtn);

        JFXTextField categoryNameField = find("#inputDialogTextField");
        categoryNameField.setText("");
        clickOn(categoryNameField);
        write(category.getName());

        confirmDialog();

        sleep(MEDIUM_SLEEP);
    }

    void addVariableTransaction(VariableTransaction transaction) {
        clickOn((JFXButton) find("#transactionsTabBtn"));
        sleep(MEDIUM_SLEEP);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);

        clickOn((JFXButton) find("#newTransactionBtn"));
        DoubleField amountTextField = find("#amountTextField");
        clickOn(amountTextField);
        eraseText(3);
        write(Double.toString(transaction.getAmount().getAmount()));
        clickOn((JFXComboBox) find("#categoryComboBox"));
        press(KeyCode.DOWN).release(KeyCode.DOWN);
        press(KeyCode.ENTER).release(KeyCode.ENTER);
        clickOn((JFXTextField) find("#purposeTextField"));
        write(transaction.getDescription());
        clickOn((JFXTextField) find("#shopTextField"));
        write(transaction.getVendor());
        DatePicker valueDatePicker = find("#valueDatePicker");
        valueDatePicker.setValue(transaction.getValueDate().getDate());

        confirmDialog();

        sleep(MEDIUM_SLEEP);
        transaction.getCategory().setId(((Category) TreeUtil.getByValue(LocalStorageImpl.getInstance().readObject("categories"),
                transaction.getCategory(), Comparator.comparing(Category::getName))).getId());
    }

    void addFixedTransaction(FixedTransaction fixedTransaction) {
        clickOn((JFXButton) find("#transactionsTabBtn"));
        sleep(MEDIUM_SLEEP);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);
        press(KeyCode.RIGHT).release(KeyCode.RIGHT);

        clickOn(find((Label label) -> label.getText().contains(fixedTransaction.getCategory().getName())));

        clickOn((JFXButton) find("#newFixedTransactionBtn"));
        sleep(SHORT_SLEEP);
        clickOn((IntegerField) find("#dayTextField"));
        eraseText(1);
        write(Integer.toString(fixedTransaction.getDay()));
        DatePicker datePicker = find("#startDateDatePicker");
        datePicker.setValue(fixedTransaction.getTimeRange().getStartDate());
        if (fixedTransaction.getIsVariable()) {
            clickOn((JFXCheckBox) find("#isVariableCheckbox"));

            sleep(100);

            for (FixedTransactionAmount transactionAmount : fixedTransaction.getTransactionAmounts()) {
                clickOn((JFXButton) find("#newTransactionAmountBtn"));

                ((DatePicker) find("#transactionAmountValueDatePicker")).setValue(transactionAmount.getValueDate().getDate());
                clickOn((DoubleField) find("#transactionAmountTextField"));
                eraseText(3);
                write(Double.toString(transactionAmount.getAmount().getAmount()));
                press(KeyCode.TAB).release(KeyCode.TAB);
                press(KeyCode.ENTER).release(KeyCode.ENTER);
            }
        } else {
            clickOn((DoubleField) find("#amountTextField"));
            eraseText(3);
            write(Double.toString(fixedTransaction.getAmount().getAmount()));
        }
        clickOn((JFXTextField) find("#productTextField"));
        write(fixedTransaction.getProduct());
        clickOn((JFXTextField) find("#purposeTextField"));
        write(fixedTransaction.getDescription());

        confirmDialog();

        sleep(SHORT_SLEEP);
    }

    final Category getCategoryTree() {
        return variableCategory();
    }

    final void confirmDialog() {
        clickOn("Ok");
    }

    @AfterEach
    void tearDownEach() throws TimeoutException {
        /* Close the window. It will be re-opened at the next test. */
        FxToolkit.hideStage();
        release(new KeyCode[]{});
        release(new MouseButton[]{});
    }
}

File: java/org.financer.client.javafx/src/test/kotlin/org/financer/client/javafx/main/AbstractFinancerApplicationTest.kt
package org.financer.client.javafx.main

import com.jfoenix.controls.*
import javafx.scene.Node
import javafx.scene.control.Button
import javafx.scene.control.Label
import javafx.scene.control.TextField
import javafx.scene.input.KeyCode
import javafx.scene.input.MouseButton
import org.financer.client.domain.model.category.Category
import org.financer.client.domain.model.transaction.*
import org.financer.client.domain.model.user.Token
import org.financer.client.domain.model.user.User
import org.financer.client.domain.model.user.VerificationToken
import org.financer.client.format.I18N.get
import org.financer.client.javafx.components.DatePicker
import org.financer.client.javafx.components.DoubleField
import org.financer.client.javafx.components.IntegerField
import org.financer.client.javafx.format.JavaFXFormatter
import org.financer.client.javafx.local.LocalStorageImpl
import org.financer.shared.domain.model.value.objects.*
import org.financer.util.collections.TreeUtil
import org.junit.jupiter.api.AfterEach
import org.junit.jupiter.api.BeforeEach
import org.testfx.api.FxToolkit
import org.testfx.framework.junit5.ApplicationTest
import java.time.LocalDate
import java.util.*
import java.util.concurrent.TimeoutException
import java.util.function.Predicate

abstract class AbstractFinancerApplicationTest : ApplicationTest() {
    var formatter: JavaFXFormatter? = null

    companion object {
        const val SHORT_SLEEP = 500
        const val MEDIUM_SLEEP = 1000
        private const val LONG_SLEEP = 5000

        init {
            System.setProperty("testfx.robot", "glass")
            System.setProperty("glass.platform", "Monocle")
            System.setProperty("monocle.platform", "Headless")
        }
    }

    protected fun user(): User {
        return User()
                .setId(1)
                .setEmail(Email("test@test.com"))
                .setName(Name("Test", "User"))
                .setPassword(HashedPassword(password()))
                .setTokens(HashSet(listOf(token())))
                .setVerified(false)
    }

    protected fun password(): String {
        return "password"
    }

    protected fun token(): Token {
        return Token()
                .setId(1)
                .setExpireDate(ExpireDate())
                .setIpAddress(IPAddress("192.168.0.1"))
                .setToken(tokenString())
                .setOperatingSystem(OperatingSystem(OperatingSystem.Values.LINUX))
    }

    protected fun verificationToken(): VerificationToken {
        return VerificationToken()
                .setId(1)
                .setExpireDate(ExpireDate(LocalDate.now().plusDays(15)))
                .setToken(tokenString())
    }

    protected fun tokenString(): TokenString {
        return TokenString("Z6XCS3tyyBlhPfsv7rMxLgfdyEOlUkv0CWSdbFYHCNX2wLlwpH97lJNP69Bny3XZ")
    }

    protected fun variableCategory(): Category {
        return Category()
                .setId(1)
                .setUser(user())
                .setCategoryClass(CategoryClass(CategoryClass.Values.VARIABLE_EXPENSES))
                .setName("Variable Category")
                .setParent(null)
    }

    protected fun variableCategoryParent(): Category {
        return Category()
                .setId(2)
                .setUser(user())
                .setCategoryClass(CategoryClass(CategoryClass.Values.VARIABLE_EXPENSES))
                .setName("Variable Category Parent")
                .setParent(null)
    }

    protected fun fixedCategory(): Category {
        return Category()
                .setId(2)
                .setUser(user())
                .setCategoryClass(CategoryClass(CategoryClass.Values.FIXED_EXPENSES))
                .setName("Fixed Category")
                .setParent(null)
    }

    protected fun product(): Product {
        return Product()
                .setId(1)
                .setName("Test Product")
                .setQuantity(Quantity(2))
                .setAmount(Amount(50))
    }

    protected fun variableTransaction(): VariableTransaction {
        return VariableTransaction()
                .setId(1)
                .setValueDate(ValueDate())
                .setCategory(variableCategory())
                .setDescription("Test Purpose")
                .setVendor("Test Vendor")
                .addProduct(product())
    }

    protected fun attachment(): Attachment {
        return Attachment()
                .setId(1)
                .setName("test.pdf")
                .setTransaction(variableTransaction())
                .setUploadDate(LocalDate.now())
    }

    protected fun fixedTransactionAmount(): FixedTransactionAmount {
        return FixedTransactionAmount()
                .setId(1)
                .setAmount(Amount(50))
                .setValueDate(ValueDate())
    }

    protected fun fixedTransaction(): FixedTransaction {
        return FixedTransaction()
                .setId(2)
                .setCategory(fixedCategory())
                .setTimeRange(TimeRange())
                .setIsVariable(false)
                .setAmount(Amount(50.0))
                .setDay(1)
                .setDescription("Test Purpose")
                .setVendor("Test Vendor")
                .addFixedTransactionAmount(fixedTransactionAmount())
    }

    @BeforeEach
    @Throws(Exception::class)
    open fun setUpEach() {
    }

    fun <T : Node?> find(query: String?): T {
        return lookup(query).query()
    }

    fun <T : Node?> find(predicate: Predicate<T>?): T {
        return lookup(predicate).query()
    }

    fun register(user: User, password: String?) {
        clickOn(find<Node>("#openRegisterDialogBtn") as JFXButton)
        clickOn(find<Node>("#registerNameTextField") as JFXTextField)
        write(user.name!!.firstName)
        clickOn(find<Node>("#registerSurnameTextField") as JFXTextField)
        write(user.name!!.surname)
        clickOn(find<Node>("#registerEmailTextField") as JFXTextField)
        write(user.email!!.emailAddress)
        val birthDatePicker = find<DatePicker>("#registerBirthDatePicker")
        birthDatePicker.value = user.birthDate!!.birthDate
        clickOn(find<Node>("#genderComboBox") as JFXComboBox<*>)
        press(KeyCode.DOWN).release(KeyCode.DOWN)
        press(KeyCode.ENTER).release(KeyCode.ENTER)
        clickOn(find<Node>("#registerPasswordTextField") as JFXPasswordField)
        write(password)
        clickOn(find<Node>("#registerRepeatPasswordTextField") as JFXPasswordField)
        write(password)
        confirmDialog()
        sleep(LONG_SLEEP.toLong())
        formatter = JavaFXFormatter(LocalStorageImpl.getInstance())
    }

    fun login(user: User, password: String?) {
        clickOn(find<Node>("#loginEmailTextField") as TextField)
        write(user.email!!.emailAddress)
        clickOn(find<Node>("#loginPasswordField") as TextField)
        write(password)
        clickOn(find<Node>("#loginBtn") as Button)
    }

    @Throws(Exception::class)
    fun logout() {
        clickOn(formatter!!.format(user().name!!))
        clickOn("Logout")
    }

    fun addCategory(category: Category) {
        clickOn(find<Node>("#profileTabBtn") as JFXButton)
        sleep(MEDIUM_SLEEP.toLong())
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        val newCategoryBtn = find<Button>("#newCategoryBtn")
        clickOn(get(category.categoryClass!!.categoryClass.getName()))
        clickOn(newCategoryBtn)
        val categoryNameField = find<JFXTextField>("#inputDialogTextField")
        categoryNameField.text = ""
        clickOn(categoryNameField)
        write(category.name)
        confirmDialog()
        sleep(MEDIUM_SLEEP.toLong())
    }

    fun addVariableTransaction(transaction: VariableTransaction) {
        clickOn(find<Node>("#transactionsTabBtn") as JFXButton)
        sleep(MEDIUM_SLEEP.toLong())
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        clickOn(find<Node>("#newTransactionBtn") as JFXButton)
        val amountTextField = find<DoubleField>("#amountTextField")
        clickOn(amountTextField)
        eraseText(3)
        write(java.lang.Double.toString(transaction.amount.amount))
        clickOn(find<Node>("#categoryComboBox") as JFXComboBox<*>)
        press(KeyCode.DOWN).release(KeyCode.DOWN)
        press(KeyCode.ENTER).release(KeyCode.ENTER)
        clickOn(find<Node>("#purposeTextField") as JFXTextField)
        write(transaction.description)
        clickOn(find<Node>("#shopTextField") as JFXTextField)
        write(transaction.vendor)
        val valueDatePicker = find<DatePicker>("#valueDatePicker")
        valueDatePicker.value = transaction.valueDate!!.date
        confirmDialog()
        sleep(MEDIUM_SLEEP.toLong())
        transaction.category!!.setId((TreeUtil.getByValue(LocalStorageImpl.getInstance().readObject("categories"),
                transaction.category, Comparator.comparing(Category::name)) as Category).id)
    }

    fun addFixedTransaction(fixedTransaction: FixedTransaction) {
        clickOn(find<Node>("#transactionsTabBtn") as JFXButton)
        sleep(MEDIUM_SLEEP.toLong())
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        press(KeyCode.RIGHT).release(KeyCode.RIGHT)
        clickOn(find(Predicate { label: Label -> label.text.contains(fixedTransaction.category!!.name!!) }))
        clickOn(find<Node>("#newFixedTransactionBtn") as JFXButton)
        sleep(SHORT_SLEEP.toLong())
        clickOn(find<Node>("#dayTextField") as IntegerField)
        eraseText(1)
        write(Integer.toString(fixedTransaction.day))
        val datePicker = find<DatePicker>("#startDateDatePicker")
        datePicker.value = fixedTransaction.timeRange!!.startDate
        if (fixedTransaction.isVariable) {
            clickOn(find<Node>("#isVariableCheckbox") as JFXCheckBox)
            sleep(100)
            for (transactionAmount in fixedTransaction.getTransactionAmounts()) {
                clickOn(find<Node>("#newTransactionAmountBtn") as JFXButton)
                (find<Node>("#transactionAmountValueDatePicker") as DatePicker).value = transactionAmount.valueDate!!.date
                clickOn(find<Node>("#transactionAmountTextField") as DoubleField)
                eraseText(3)
                write(java.lang.Double.toString(transactionAmount.amount.amount))
                press(KeyCode.TAB).release(KeyCode.TAB)
                press(KeyCode.ENTER).release(KeyCode.ENTER)
            }
        } else {
            clickOn(find<Node>("#amountTextField") as DoubleField)
            eraseText(3)
            write(java.lang.Double.toString(fixedTransaction.amount.amount))
        }
        clickOn(find<Node>("#productTextField") as JFXTextField)
        write(fixedTransaction.product)
        clickOn(find<Node>("#purposeTextField") as JFXTextField)
        write(fixedTransaction.description)
        confirmDialog()
        sleep(SHORT_SLEEP.toLong())
    }

    val categoryTree: Category
        get() = variableCategory()

    fun confirmDialog() {
        clickOn("Ok")
    }

    @AfterEach
    @Throws(TimeoutException::class)
    open fun tearDownEach() {
        /* Close the window. It will be re-opened at the next test. */
        FxToolkit.hideStage()
        release(*arrayOf<KeyCode>())
        release(*arrayOf<MouseButton>())
    }
}